File: meteo-backend/app/models/weather.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field


class ForecastPoint(BaseModel):
    datetime: datetime
    temperature_c: Optional[float] = None
    humidity_percent: Optional[float] = Field(default=None, ge=0, le=100)
    precipitation_probability: Optional[float] = Field(default=None, ge=0, le=100)
    precipitation_total: Optional[float] = Field(default=None, ge=0)
    precipitation_snow: Optional[float] = Field(default=None, ge=0)
    cloud_cover: Optional[float] = Field(default=None, ge=0)
    wind_speed_kmh: Optional[float] = Field(default=None, ge=0)
    apparent_temperature_c: Optional[float] = Field(default=None)


class ProviderForecast(BaseModel):
    provider: str
    latitude: float
    longitude: float
    timezone: str
    forecast: list[ForecastPoint]


class AggregatedStat(BaseModel):
    min: Optional[float] = None
    avg: Optional[float] = None
    max: Optional[float] = None


class AggregatedHourlyForecastPoint(BaseModel):
    datetime: datetime
    provider_count: int = Field(ge=0)
    temperature_c: AggregatedStat = Field(default_factory=AggregatedStat)
    precipitation_probability: AggregatedStat = Field(default_factory=AggregatedStat)
    precipitation_total: AggregatedStat = Field(default_factory=AggregatedStat)
    precipitation_snow: AggregatedStat = Field(default_factory=AggregatedStat)
    humidity_percent: Optional[float] = None
    cloud_cover: Optional[float] = None
    wind_speed_kmh: Optional[float] = None
    apparent_temperature_c: Optional[float] = None
    condition: str


class AggregatedDailyForecastPoint(BaseModel):
    date: date
    provider_count: int = Field(ge=0)
    temperature_min_c: AggregatedStat = Field(default_factory=AggregatedStat)
    temperature_max_c: AggregatedStat = Field(default_factory=AggregatedStat)
    precipitation_total: AggregatedStat = Field(default_factory=AggregatedStat)
    condition: str


class AggregationWindow(BaseModel):
    mode: str
    start: Optional[datetime] = None
    end: Optional[datetime] = None


class DailyAggregationWindow(BaseModel):
    mode: str
    start: Optional[date] = None
    end: Optional[date] = None


class AggregatedForecast(BaseModel):
    latitude: float
    longitude: float
    timezone: str
    days: int = Field(ge=1)
    providers_requested: list[str]
    providers_used: list[str]
    provider_errors: dict[str, str] = Field(default_factory=dict)
    warnings: list[str] = Field(default_factory=list)
    hourly_window: AggregationWindow
    daily_window: DailyAggregationWindow
    hourly_forecast: list[AggregatedHourlyForecastPoint]
    daily_forecast: list[AggregatedDailyForecastPoint]

File: frontend/src/types/weather.ts
export type WeatherCondition =
  | "sunny"
  | "partly_cloudy"
  | "cloudy"
  | "rain"
  | "snow"
  | "unknown";

export interface AggregatedStat {
  min: number | null;
  avg: number | null;
  max: number | null;
}

export interface TemperatureConsensusStat {
  consensus_low: number | null;
  central: number | null;
  consensus_high: number | null;
  provider_min: number | null;
  provider_max: number | null;
  sample_count: number;
}

export interface LocationCandidate {
  id: string;
  provider: string;
  provider_id: number;
  name: string;
  display_name: string;
  latitude: number;
  longitude: number;
  timezone: string;
  country: string | null;
  country_code: string | null;
  admin1: string | null;
  admin2: string | null;
  admin3: string | null;
  admin4: string | null;
  elevation: number | null;
  population: number | null;
}

export interface LocationSearchResponse {
  query: string;
  count: number;
  results: LocationCandidate[];
}

export interface AggregatedHourlyForecastPoint {
  datetime: string;
  provider_count: number;
  temperature_c: TemperatureConsensusStat;
  precipitation_probability: AggregatedStat;
  precipitation_total: AggregatedStat;
  precipitation_snow: AggregatedStat;
  humidity_percent: number | null;
  cloud_cover: number | null;
  wind_speed_kmh: number | null;
  apparent_temperature_c: number | null;
  condition: WeatherCondition;
}

export interface AggregatedDailyForecastPoint {
  date: string;
  provider_count: number;
  temperature_min_c: TemperatureConsensusStat;
  temperature_max_c: TemperatureConsensusStat;
  precipitation_total: AggregatedStat;
  condition: WeatherCondition;
}

export interface AggregatedForecast {
  latitude: number;
  longitude: number;
  timezone: string;
  days: number;
  providers_requested: string[];
  providers_used: string[];
  provider_errors: Record<string, string>;
  warnings: string[];
  hourly_window: {
    mode: string;
    start: string | null;
    end: string | null;
  };
  daily_window: {
    mode: string;
    start: string | null;
    end: string | null;
  };
  hourly_forecast: AggregatedHourlyForecastPoint[];
  daily_forecast: AggregatedDailyForecastPoint[];
}
